### PR TITLE
generate CLI docs automatically from command help text

### PR DESCRIPTION
**Pull request by ozkatz**
Added `lakectl docs` that spits out a markdown formatted documentation for the `lakectl` command tree. For convenience, added `make gen-docs` that updates the CLI reference page under `docs/reference/commands.md` with the output of this.

This provides 2 benefits:

1. We can automate this further to ensure there's never a documentation drift (well, for command line reference at least)
1. Creates a single source of truth for help information - the Command structs.

### CODE REVIEW

**Comment by ozkatz** [comment]
Closes #1409

File: cmd/lakectl/cmd/docs.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+			DieErr(err)
+		}
+		// gen child commands now
+		err = recurseGenMarkdown(c, w)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
```
Comment (nopcoder) [inline cmd/lakectl/cmd/docs.go:117]: Suggestion: this func should first call `genMarkdownForCmd` on cmd and all cmd.Commands().
This way the docs command just need to call this recursive function.

**Review by nopcoder — APPROVED** [review]
lgtm

**Review by johnnyaug — APPROVED** [review]

File: cmd/lakectl/cmd/docs.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+	buf.WriteString("\n\n")
+	_, err := buf.WriteTo(w)
+	if err != nil {
+		return err
+	}
+
+	// recurse to children
+	for _, c := range cmd.Commands() {
+		err := genMarkdownForCmd(c, w)
```
Comment (arielshaqed) [inline cmd/lakectl/cmd/docs.go:107]:
```suggestion
		if c.Hidden { continue }
		err := genMarkdownForCmd(c, w)
```

File: cmd/lakectl/cmd/docs.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+will provide a link to download a preconfigured configuration file for you.
+
+`
+
+func printOptions(buf *bytes.Buffer, cmd *cobra.Command) error {
+	flags := cmd.NonInheritedFlags()
+	flags.SetOutput(buf)
+	if flags.HasAvailableFlags() {
+		buf.WriteString("#### Options\n\n```\n")
+		flags.PrintDefaults()
```
Comment (arielshaqed) [inline cmd/lakectl/cmd/docs.go:62]: Does this skip `Hidden` flags?

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Should skip everyting "Hidden".  Not sure if there are others, but I found 2 -- one of which needs actual code AFAICS.

File: cmd/lakectl/cmd/docs.go (hunk 3 above)
Comment (nopcoder) [reply cmd/lakectl/cmd/docs.go:62]: // HasAvailableFlags returns a bool to indicate if the FlagSet has any flags
// that are not hidden.

Comment (arielshaqed) [reply cmd/lakectl/cmd/docs.go:62]: OK, cool.  So actually `PrintDefaults` [skips all hidden flags](https://github.com/spf13/pflag/blob/85dd5c8bc61cfa382fecd072378089d4e856579d/flag.go#L691).

Resolving, thanks!

File: cmd/lakectl/cmd/docs.go (hunk 2 above)
Comment (ozkatz) [reply cmd/lakectl/cmd/docs.go:112]: Added a disclaimer/warning hidden commands. Please rereview :)

**Review by arielshaqed — APPROVED** [review]
Great; thanks!

**Pull request merged**